python/lsst/ts/externalscripts/base_make_calibrations.py

```diff
@@ -0,0 +1,640 @@
+                    raise ValueError(
+                        "n_images_" + f"{image_type}".lower() + f"={n_images} specified and "
+                        "exp_times_" + f"{image_type}".lower() + f"={exp_times} is an array, "
+                        f"but the length does not match the number of images."
+                    )
+        else:
+            # exp_times is a scalar; if n_images is specified then
+            # take that many images, else take 1 image
+            if n_images is None:
+                n_images = 1
+            exp_times = [exp_times] * n_images
+
+        return exp_times
+
+    async def configure(self, config):
+        """Configure the script.
+
+        Parameters
+        ----------
+        config: `types.SimpleNamespace`
+            Configuration data. See `get_schema` for information about data
+            structure.
+        """
+        # Log information about the configuration
+
+        self.log.debug(
+            f"n_bias: {config.n_bias}, detectors: {config.detectors}, "
+            f"n_dark: {config.n_dark}, "
+            f"n_flat: {config.n_flat}, "
+            f"instrument: {self.instrument_name} "
+        )
+
+        self.config = config
+
+    def set_metadata(self, metadata):
+        """Set estimated duration of the script.
+        """
+        metadata.duration = self.config.n_bias*(self.camera.read_out_time + self.estimated_process_time)
+
+    async def take_image_type(self, image_type, exp_times):
+        """Take exposures and build exposure set.
+
+        Parameters
+        ----------
+        image_type : `str`
+            Image type. One of ["BIAS", "DARK", "FLAT"].
+
+        exp_times: `list`
+            List of exposure times.
+
+        Returns
+        -------
+            Tuple with exposure IDs.
+        """
+
+        return tuple(
+            [
+                (await self.camera.take_imgtype(image_type, exp_time, 1))[0]
+                for exp_time in exp_times
+            ]
+        )
+
+    async def image_in_oods_callback(self, data):
+        """Callback function to check images are in archiver"""
+        self.image_in_oods_samples[self.current_image_type].append(data)
+        self.number_of_images_taken += 1
+        if self.number_of_images_taken == self.number_of_images_expected:
+            self.image_in_oods_received_all_expected.set()
+
+    async def take_images(self, image_type):
+        """Take images with instrument.
+
+        Parameters
+        ----------
+        image_type : `str`
+            Image type. One of ["BIAS", "DARK", "FLAT"].
+
+        Returns
+        -------
+        exposures : `tuple`
+             Tuple with the IDs of the exposures taken.
+        """
+
+        exp_times = await self.set_exp_times_per_im_type(image_type)
+
+        n_detectors = len(tuple(map(int, self.config.detectors[1:-1].split(','))))
+
+        self.number_of_images_expected = len(exp_times)*n_detectors
+        self.number_of_images_taken = 0
```

Comment:
Is there a check that `number_of_images_taken == number_of_images_expected`?  I don't think I see one here, or below when this is called.

Reply:
I think that's done in the function ``image_in_oods_callback`` define in L313.  it increases a counter every time an image is taken, and then sets ``image_in_oods_received_all_expected`` when the condition is satisfied (i.e., taken == expected).

Reply:
The `image_in_oods_callback` has the logic to handle this condition. Basically if we get less images, the line below:

```
await asyncio.wait_for(self.image_in_oods_received_all_expected.wait(),
                                timeout=self.config.oods_timeout)
```

will timeout. Now, there could be a situation where we get more images than we expect, which probably means someone else commanded the camera to take images. This could be a fail condition which we check afterwards.